apps/desktop-wallet/src/api/addresses.ts

```diff
@@ -114,3 +115,117 @@ export const fetchAddressesTransactionsNextPage = async (addresses: Address[], n
 
   return transactions
 }
+
+export const discoverAndCacheActiveAddresses = async (
+  addressIndexesToSkip: number[] = [],
+  minGap = 5
+): Promise<NonSensitiveAddressData[]> => {
+  const addressesPerGroup = Array.from({ length: TOTAL_NUMBER_OF_GROUPS }, (): NonSensitiveAddressData[] => [])
+  const activeAddresses: NonSensitiveAddressData[] = []
+  const skipIndexes = Array.from(addressIndexesToSkip)
+
+  for (let group = 0; group < TOTAL_NUMBER_OF_GROUPS; group++) {
+    const newAddresses = deriveAddressesInGroup(group, minGap, skipIndexes)
+    addressesPerGroup[group] = newAddresses
+    skipIndexes.push(...newAddresses.map((address) => address.index))
+  }
+
+  const addressesToCheckIfActive = addressesPerGroup.flat().map((address) => address.hash)
+  const results = await getActiveAddressesResults(addressesToCheckIfActive)
+  const resultsPerGroup = splitResultsArrayIntoOneArrayPerGroup(results, minGap)
+
+  for (let group = 0; group < TOTAL_NUMBER_OF_GROUPS; group++) {
+    const { gap, activeAddresses: newActiveAddresses } = getGapFromLastActiveAddress(
+      addressesPerGroup[group],
+      resultsPerGroup[group]
+    )
+
+    let gapPerGroup = gap
+    activeAddresses.push(...newActiveAddresses)
+
+    while (gapPerGroup < minGap) {
+      const remainingGap = minGap - gapPerGroup
+      const newAddresses = deriveAddressesInGroup(group, remainingGap, skipIndexes)
+      skipIndexes.push(...newAddresses.map((address) => address.index))
+
+      const newAddressesToCheckIfActive = newAddresses.map((address) => address.hash)
+      const results = await getActiveAddressesResults(newAddressesToCheckIfActive)
+
+      const { gap, activeAddresses: newActiveAddresses } = getGapFromLastActiveAddress(
+        newAddresses,
+        results,
+        gapPerGroup
+      )
+      gapPerGroup = gap
+      activeAddresses.push(...newActiveAddresses)
+    }
+  }
+
+  return activeAddresses
+}
+
+const deriveAddressesInGroup = (group: number, amount: number, skipIndexes: number[]): NonSensitiveAddressData[] => {
```

Comment:
This function along with `splitResultsArrayIntoOneArrayPerGroup` could go in the `utils/addresses.ts` file, so that we keep only api related async functions in here.